README.md
# nsfp

A command-line player of Nintendo NES/Famicom music files (.nsf/.nsfe)

Based on the Game Music Emu library.

## Usage

```
$ nsfp Kirby.nes
nsfp 0.1 - Simple command-line player of NSF/NSFE files
Game:      Kirby's Adventure
Author:    Hirokazu Ando, Jun Ishikawa
Copyright: 1993 HAL Laboratory, Nintendo

Kirby's Adventure: 0/58  (2:30)
...
```

## Dependencies

On Debian/Ubuntu you need a recent GCC compiler, and the development files
forSDL 2.0 and Game Music Emu:

```
sudo apt-get install build-essential libsdl2-dev libgme-dev
```

## Install

After cloning the repository, go to the directory and run the following to
compile and install the player:

```
mkdir build && cd build
cmake ..
make
sudo make install
```

## License

Source code is released under Apache 2.0 license. Please refer to
[LICENSE](LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

* Basic playback of NSF files
* Show track info


src/player.h
/*
 * Copyright 2018 Damián Silvani
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef __PLAYER_H__
#define __PLAYER_H__

#include "gme/gme.h"
#include <string>

typedef short sample_t;

class Player {
public:
  Player();
  ~Player();

  // Initialize player and set sample rate
  gme_err_t init(long sample_rate = 44100);

  // Load game music file. NULL on success, otherwise error string.
  gme_err_t load_file(const std::string &path);

  // (Re)start playing track. Tracks are numbered from 0 to track_count() - 1.
  gme_err_t start_track(int track);

  // Stop playing current file
  void stop();

  //
  // Optional functions
  //

  // Return currently loaded filename
  const std::string &filename() { return filename_; }

  // Number of tracks in current file, or 0 if no file loaded.
  int track_count() const;

  // Info for current track
  gme_info_t const &track_info() const { return *track_info_; }

  // Pause/resume playing current track.
  void pause(int);

  // True if track ended
  bool track_ended() const;

  // Pointer to emulator
  Music_Emu *emu() const { return emu_; }

  // Set stereo depth, where 0.0 = none and 1.0 = maximum
  void set_stereo_depth(double);

  // Enable accurate sound emulation
  void enable_accuracy(bool);

  // Set tempo, where 0.5 = half speed, 1.0 = normal, 2.0 = double speed
  void set_tempo(double);

  // Set voice muting bitmask
  void mute_voices(int);

private:
  Music_Emu *emu_;
  long sample_rate;
  bool paused;
  gme_info_t *track_info_;
  std::string filename_;

  void suspend();
  void resume();
  static void fill_buffer(void *, sample_t *, int);
};

#endif // __PLAYER_H__


src/player.cc
/*
 * Copyright 2018 Damián Silvani
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "player.h"
#include <cstring>
#include <string>

using namespace std;

#define RETURN_ERR(expr)                                                       \
  do {                                                                         \
    gme_err_t err_ = (expr);                                                   \
    if (err_)                                                                  \
      return err_;                                                             \
  } while (0)

// Number of audio buffers per second. Adjust if you encounter audio skipping.
const int fill_rate = 45;

// Simple sound driver using SDL
typedef void (*sound_callback_t)(void *data, short *out, int count);
static const char *sound_init(long sample_rate, int buf_size, sound_callback_t,
                              void *data);
static void sound_start();
static void sound_stop();
static void sound_cleanup();

Player::Player() {
  emu_ = 0;
  paused = false;
  track_info_ = nullptr;
}

gme_err_t Player::init(long rate) {
  sample_rate = rate;

  int min_size = sample_rate * 2 / fill_rate;
  int buf_size = 512;
  while (buf_size < min_size)
    buf_size *= 2;

  return sound_init(sample_rate, buf_size, fill_buffer, this);
}

void Player::stop() {
  sound_stop();
  gme_delete(emu_);
  emu_ = nullptr;
}

Player::~Player() {
  stop();
  sound_cleanup();
  gme_free_info(track_info_);
}

gme_err_t Player::load_file(const string &path) {
  filename_ = path;

  stop();

  RETURN_ERR(gme_open_file(path.c_str(), &emu_, sample_rate));

  char m3u_path[256 + 5];
  strncpy(m3u_path, path.c_str(), 256);
  m3u_path[256] = 0;
  char *p = strrchr(m3u_path, '.');
  if (!p)
    p = m3u_path + strlen(m3u_path);
  strcpy(p, ".m3u");
  if (gme_load_m3u(emu_, m3u_path)) {
  } // ignore error

  return 0;
}

int Player::track_count() const { return emu_ ? gme_track_count(emu_) : false; }

gme_err_t Player::start_track(int track) {
  if (emu_) {
    gme_free_info(track_info_);
    track_info_ = nullptr;
    RETURN_ERR(gme_track_info(emu_, &track_info_, track));

    // Sound must not be running when operating on emulator
    sound_stop();
    RETURN_ERR(gme_start_track(emu_, track));

    // Calculate track length
    if (track_info_->length <= 0)
      track_info_->length =
          track_info_->intro_length + track_info_->loop_length * 2;

    if (track_info_->length <= 0)
      track_info_->length = (long)(2.5 * 60 * 1000);
    gme_set_fade(emu_, track_info_->length);

    paused = false;
    sound_start();
  }
  return 0;
}

void Player::pause(int b) {
  paused = b;
  if (b)
    sound_stop();
  else
    sound_start();
}

void Player::suspend() {
  if (!paused)
    sound_stop();
}

void Player::resume() {
  if (!paused)
    sound_start();
}

bool Player::track_ended() const {
  return emu_ ? gme_track_ended(emu_) : false;
}

void Player::set_stereo_depth(double tempo) {
  suspend();
  gme_set_stereo_depth(emu_, tempo);
  resume();
}

void Player::enable_accuracy(bool b) {
  suspend();
  gme_enable_accuracy(emu_, b);
  resume();
}

void Player::set_tempo(double tempo) {
  suspend();
  gme_set_tempo(emu_, tempo);
  resume();
}

void Player::mute_voices(int mask) {
  suspend();
  gme_mute_voices(emu_, mask);
  gme_ignore_silence(emu_, mask != 0);
  resume();
}

void Player::fill_buffer(void *data, sample_t *out, int count) {
  Player *self = (Player *)data;
  if (self->emu_) {
    if (gme_play(self->emu_, count, out)) {
    } // ignore error
  }
}

// Sound output driver using SDL

#include "SDL2/SDL.h"

static sound_callback_t sound_callback;
static void *sound_callback_data;

static void sdl_callback(void *data, Uint8 *out, int count) {
  if (data) {
  }; // ignore unused variable warning
  if (sound_callback)
    sound_callback(sound_callback_data, (short *)out, count / 2);
}

static const char *sound_init(long sample_rate, int buf_size,
                              sound_callback_t cb, void *data) {
  sound_callback = cb;
  sound_callback_data = data;

  static SDL_AudioSpec as; // making static clears all fields to 0
  as.freq = sample_rate;
  as.format = AUDIO_S16SYS;
  as.channels = 2;
  as.callback = sdl_callback;
  as.samples = buf_size;
  if (SDL_OpenAudio(&as, 0) < 0) {
    const char *err = SDL_GetError();
    if (!err)
      err = "Couldn't open SDL audio";
    return err;
  }

  return 0;
}

static void sound_start() { SDL_PauseAudio(false); }

static void sound_stop() {
  SDL_PauseAudio(true);

  // be sure audio thread is not active
  SDL_LockAudio();
  SDL_UnlockAudio();
}

static void sound_cleanup() {
  sound_stop();
  SDL_CloseAudio();
}


src/main.cc
/*
 * Copyright 2018 Damián Silvani
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "player.h"
#include <SDL2/SDL.h>
#include <iostream>
#include <string>

using namespace std;

void start_track(Player *player, int track, bool print_full_info = false) {
  // Start first track
  if (auto err = player->start_track(track)) {
    cerr << "Player error: " << err << endl;
    exit(1);
  }

  // Get track information
  long seconds = player->track_info().length / 1000;
  const char *game = player->track_info().game;
  if (!*game) {
    // extract filename
    game = strrchr(player->filename().c_str(), '\\'); // DOS
    if (!game)
      game = strrchr(player->filename().c_str(), '/'); // UNIX
    if (!game)
      game = player->filename().c_str();
    else
      game++; // skip path separator
  }

  auto &info = player->track_info();

  if (print_full_info) {
    if (strcmp(info.game, "") != 0)
      cout << "Game:      " << info.game << "\n";
    if (strcmp(info.author, "") != 0)
      cout << "Author:    " << info.author << "\n";
    if (strcmp(info.copyright, "") != 0)
      cout << "Copyright: " << info.copyright << "\n";
    if (strcmp(info.comment, "") != 0)
      cout << "Comment:   " << info.comment << "\n";
    if (strcmp(info.dumper, "") != 0)
      cout << "Dumper:    " << info.dumper;
    cout << endl;
  }

  char title[512];
  sprintf(title, "%s: %d/%d %s (%ld:%02ld)", game, track, player->track_count(),
          player->track_info().song, seconds / 60, seconds % 60);

  cout << title << endl;
}

int main(int argc, char *argv[]) {
  cerr << "nsfp 0.1 - Simple command-line player of NSF/NSFE files" << endl;

  if (argc != 2) {
    cerr << "Usage: " << argv[0] << " nsf_file" << endl;
    return 1;
  }

  const string filename = string(argv[1]);

  if (SDL_Init(SDL_INIT_AUDIO) < 0) {
    cerr << "Failed to initialize SDL" << endl;
    return 1;
  }
  atexit(SDL_Quit);

  // Create player
  Player *player = new Player;
  if (!player) {
    cerr << "Out of memory" << endl;
    return 1;
  }

  // Initialize
  if (auto err = player->init()) {
    cerr << "Player error: " << err << endl;
    return 1;
  }

  // Load file
  if (auto err = player->load_file(filename)) {
    cerr << "Player error: " << err << endl;
    return 1;
  }

  bool running = true;

  int track = 0;
  start_track(player, track, true);

  while (running) {
    SDL_Delay(1000);
  }

  delete player;

  return 0;
}
